tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -488,6 +516,164 @@ static void read_kobject_user_pipe(void)
 		"did not fault\n");
 }
 
+static void high_prio_uthread(void)
+{
+	u8_t new_prio;
+
+	expect_fault = true;
+	expected_reason = REASON_KERNEL_OOPS;
+	BARRIER();
+	/* Get the priority of current thread and get higher
+	 * priority value
+	 */
+	new_prio = k_thread_priority_get(k_current_get()) - 2;
+
+	/* Create a user thread with priority greater than current
+	 * thread
+	 */
+	k_thread_create(&uthread_thread, uthread_stack, STACKSIZE,
+			(k_thread_entry_t)thread_body, NULL, NULL, NULL,
+			K_PRIO_PREEMPT(new_prio), K_USER | K_INHERIT_PERMS,
+			K_NO_WAIT);
+
+	/* This part of the code should not be reached */
+	zassert_unreachable("Create higher priority user thread greater"
+			" than current thread did not fault\n");
+}
+
+struct k_mem_domain app0_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app0_buf0[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT app0_buf1[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(app0_part0, app0_buf0, sizeof(app0_buf0),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+K_MEM_PARTITION_DEFINE(app0_part1, app0_buf1, sizeof(app0_buf1),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *app0_parts[] = {
+	&app0_part0,
+	&app0_part1
+};
+
+struct k_mem_domain app1_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app1_buf0[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT app1_buf1[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(app1_part0, app1_buf0, sizeof(app1_buf0),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+K_MEM_PARTITION_DEFINE(app1_part1, app1_buf1, sizeof(app1_buf1),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *app1_parts[] = {
+	&app1_part0,
+	&app1_part1
+};
+
+__kernel struct k_thread utid, ktid;
+K_THREAD_STACK_DEFINE(ustack, MEM_DOMAIN_STACK_SIZE);
+K_THREAD_STACK_DEFINE(kstack, MEM_DOMAIN_STACK_SIZE);
+
+/***************************************************************************/
+static void kthread_entry(void *p1, void *p2, void *p3)
+{
+	/* Add current thread to memory domain app0_domain*/
+	k_mem_domain_add_thread(&app0_domain, k_current_get());
+
+	/* Add current thread to another memory domain app1_domain
+	 * This should cause a assertion failure, provided CONFIG_ASSERT=y
+	 * Cannot add a thread to two memory domains
+	 */
+	k_mem_domain_add_thread(&app1_domain, k_current_get());
+
+		/* These lines of code should not be reached */
+	zassert_unreachable("Adding a thread to multiple memory domains"
+			" and accessing them did not fault\n");
+}
+
+/* Verify that a thread can belong to a single memory domain
+ * at any point of time
+ * Try to add a thread to two memory domains at a time
+ */
+static void thread_in_multiple_domains(void)
+{
+	k_mem_domain_init(&app0_domain, ARRAY_SIZE(app0_parts), app0_parts);
+	k_mem_domain_init(&app1_domain, ARRAY_SIZE(app1_parts), app1_parts);
+
+	k_thread_create(&kthread_thread, kthread_stack, STACKSIZE,
+			(k_thread_entry_t)kthread_entry, NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1), 0, K_NO_WAIT);
+
+	k_sem_take(&sync_sem, SEM_TIMEOUT);
+}
```

Comment:
This is not a test of userspace AFAICT; it is testing kernel assertion checking on kernel threads performing memory domain setup.  Probably belongs in a separate test program? Also, I am not a fan of this k_sem_take(&sync_sem, SEM_TIMEOUT) idiom for tests; it seems unreliable (could timeout and yield a false pass result).  We are using the approach k_thread_abort(k_current_get()) in the ztest thread to stop the ztest thread from generating a pass or fail, and instead handling both in the newly created thread.

Reply:
See revoke_other_thread() for an example of this approach.

Reply:
Yes, I agree that this belong to separate set of tests. This test case will be merged with #5197. In this test, ASSERT failure is hit and the thread goes to infinite loop. Hence calling k_sem_take(&sync_sem, SEM_TIMEOUT) in the main thread. I will add a condition to check from ztest to explicitly know which part of the ASSERT is failed.